Src/Idioms/EmptyStringBehaviorExpectation.cs

```diff
@@ -0,0 +1,49 @@
+﻿using System;
+
+namespace AutoFixture.Idioms
+{
+    /// <summary>
+    /// Encapsulates the expected behavior when an <see cref="IGuardClauseCommand" /> (typically
+    /// representing a method or constructor) is invoked with a <see cref="string.Empty" /> argument.
+    /// </summary>
+    /// <seealso cref="Verify(IGuardClauseCommand)" />
+    public class EmptyStringBehaviorExpectation : IBehaviorExpectation
+    {
+        /// <summary>
+        /// Verifies that the command behaves correct when invoked with an empty string argument.
+        /// </summary>
+        /// <param name="command">The command whose behavior must be examined.</param>
+        /// <remarks>
+        /// <para>
+        /// This method encapsulates the behavior which is expected when a method or constructor is
+        /// invoked with <see cref="string.Empty" /> as one of the method arguments. In that case it's
+        /// expected that invoking <paramref name="command" /> with string.Empty throws an
+        /// <see cref="ArgumentException" />, causing the Verify method to succeed. If other
+        /// exceptions are thrown, or no exception is thrown when invoking the command, the Verify
+        /// method throws an exception.
+        /// </para>
+        /// </remarks>
+        public void Verify(IGuardClauseCommand command)
+        {
+            if (command == null) throw new ArgumentNullException(nameof(command));
+
+            if (command.RequestedType != typeof(string))
+                return;
+
+            try
+            {
+                command.Execute(string.Empty);
+            }
+            catch (ArgumentException)
```

Comment:
`NullReferenceBehaviorExpectation` uses `ArgumentNullException` which takes a parameter name in its constructor, whereas `ArgumentException` does not.  I believe this is why `EmptyGuidBehaviorExpectation` does not have the same parameter name test as `NullReferenceBehaviorExpectation`.  I could test for a message, like "The parameter name is invalid" or something like that.

Reply:
I've just tested that and found that can pass parameter name to the `ArgumentException`:
```c#
var exception = new ArgumentException("Empty string", nameof(arg));
var paramName = exception.ParamName;
```

Could you please verify for one more time whether I'm missing something?

Reply:
Clearly, my brain was operating poorly...  I made the change and added the test.

Reply:
However...  `ArgumentNullException` has an overload that allows you to call it with just a parameter name; `ArgumentException` does not.  Calling `ArgumentException` with only one parameter causes `message` to be populated, not `parameterName`.  This is problematic if a person doesn't create their Guard Clause for empty strings with both a message and a parameter name.  `EmptyGuidBehaviorExpectation` doesn't have this problem since it does not have the same parameter check as `NullReferenceBehaviorExpectation`.  In other words, with the way I wrote `EmptyStringBehaviorExpectation`, this Guard Clause in a class being verified will not work:

        private class EmptyStringTest
        {
            public EmptyStringTest(string bogus)
            {
                if (bogus == string.Empty)
                    throw new ArgumentException(nameof(bogus));
            }
        }
However, this will work for `EmptyGuidBehaviorExpectation`:

        private class EmptyGuidTest
        {
            public EmptyGuidTest(Guid bogus)
            {
                if (bogus == Guid.Empty)
                    throw new ArgumentException(nameof(bogus));
            }
        }

They look the same, but, since `EmptyGuidBehaviorExpectation` does NOT compare parameter names, it works if a Guid is empty, whereas `EmptyStringBehaviorExpectation` does not work if a string is empty since the parameter name is null.  To make my current implementation of `EmptyStringBehaviorExpectation` function properly, a developer would have to construct their Guard Clause (for empty strings) like so in order to populate the parameter name:

        private class EmptyStringTest
        {
            public EmptyStringTest(string bogus)
            {
                if (bogus == string.Empty)
                    throw new ArgumentException("my message", nameof(bogus));
            }
        }

So to me, it feels like the expected behavior between empty Guid and empty String shouldn't be different; I should probably remove the parameter check.  The other option is to add a parameter check to empty Guid, but that would be a breaking change.

Any thoughts?

Reply:
Thanks for describing your point in detail! I've reviewed it and here are my thoughts.

Personally, I feel that `EmptyGuidBehaviorExpectation` should also verify the parameter name. I believe that the reason why it doesn't do this is that we simply overlooked the inconsistency. That's something that should be fixed in v5 and [I've created](https://github.com/AutoFixture/AutoFixture/issues/1009) an issue for that.

My reasoning is that if you have the following guard (from the sample above):
```c#
if (string.IsNullOrEmpty(secondArg)) throw new ArgumentException(nameof(secondArg));
```
then the thrown exception doesn't contain any explanation of what happened:
```
System.ArgumentException
secondArg
```

That's not what you expect from the exception message as it should contain the meaningful message. Compare that to `ArgumentNullException`:
```
System.ArgumentNullException
Value cannot be null.
Parameter name: secondName
```

For me, the goal of the `GuardClauseAssertion` is to:
- verify a guard clause is present
- verify a guard clause is valid, i.e. it doesn't violate the common .NET contract.

I feel that exception without meaningful message definitely violates the .NET contract and we should help to catch such things.

Notice, you have a few ways to have nice exceptions and don't duplicate message all over the code:
- Create aux method to verify value (e.g. `Assert.ArgumentIsNotNullOrEmpty(..)`), so message is specified there.
-  Create a custom exception like `ArgumentNullException` taking the parameter name only.

Actually, the last option is preferable to me and here is how it looks:
```c#
public class ArgumentNullOrEmptyException : ArgumentException
{
    public ArgumentNullOrEmptyException(string paramName)
        : base("Value should not be null or empty.", paramName)
    {
    }
}

......
if (string.IsNullOrEmpty(secondArg)) throw new ArgumentNullOrEmptyException(nameof(secondArg));
```

Result:
```
AutoFixturePlayground.ArgumentNullOrEmptyException
Value should not be null or empty.
Parameter name: secondArg
```

A few minor changes to the code and now your log is not full of awkward errors - isn't that awesome? :)